src/types/list.rs

```diff
@@ -177,17 +177,83 @@ impl PyList {
     }
 }
 
+macro_rules! panic_out_of_range {
+    ($index:expr, $len:expr) => {
+        panic!("index {} out of range for list of length {}", $index, $len)
+    };
+}
+
 impl Index<usize> for PyList {
     type Output = PyAny;
 
     fn index(&self, index: usize) -> &Self::Output {
-        self.get_item(index).unwrap_or_else(|_| {
-            panic!(
-                "index {} out of range for list of length {}",
-                index,
-                self.len()
-            );
-        })
+        self.get_item(index)
+            .unwrap_or_else(|_| panic_out_of_range!(index, self.len()))
+    }
+}
+
+impl Index<Range<usize>> for PyList {
+    type Output = PyList;
+
+    fn index(&self, Range { start, end }: Range<usize>) -> &Self::Output {
+        let len = self.len();
+        if start >= len {
```

Comment:
Actually, Rust allows `start` to be `== len` too.

You need an additional check for `start <= end`.